blogpost/blogpost.md

```diff
@@ -0,0 +1,845 @@
+    <artifactId>beam-sdks-java-io-mongodb</artifactId>
+    <version>2.24.0</version>
+</dependency>
+```
+Same, for the ability to write to MongoDB.
+
+Right now our app contains all the needed dependencies and a simple Java class LightningApplication.
+
+### Set the Apache Kafka and MongoDB instances
+
+The Lightning app will connect to Apache Kafka and MongoDB. They can be easily set using Docker Compose:
+```yml
+version: '3.7'
+
+services:
+  zookeeper:
+    image: bitnami/zookeeper:3.6.2
+    container_name: lightning_zookeeper
+    ports:
+      - '2181:2181'
+    environment:
+      - ALLOW_ANONYMOUS_LOGIN=yes
+
+  kafka:
+    image: bitnami/kafka:2.6.0
+    container_name: lightning_kafka
+    links:
+      - zookeeper:zookeeper
+    ports:
+      - '9092:9092'
+    environment:
+      - KAFKA_CFG_ZOOKEEPER_CONNECT=zookeeper:2181
+      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://localhost:9092
+      - ALLOW_PLAINTEXT_LISTENER=yes
+    depends_on:
+      - zookeeper
+
+  mongo:
+    image: mongo:4.4.1
+    container_name: lightning_mongo
+    ports:
+      - '27017:27017'
+    environment:
+      - MONGO_INITDB_DATABASE=lightning_db
+```
+This docker-compose file sets up an Apache Kafka linked with Apache Zookeeper and a MongoDB instances.
+
+### Spring Boot app configuration
+
+Now get back to the app. Here are all the needed configurations for application.yml:
+```yml
+mongo:
+  host: mongodb://localhost:27017
+  database: lightning_db
+  collection:
+    lightnings: lightnings
+    strikes: strikes
+beam:
+  window:
+    size: 60
+
+kafka:
+  bootstrap:
+    servers: localhost:9092
+  topic: lightning_topic
+  auto:
+    offset:
+      reset: latest
+
+spring:
+  profiles:
+    active: withEmulator
+```
+Here I define mongo, beam and kafka environment properties so the app will have one source of truth. It's in Yaml file format that is way more readable than .properties files. Spring supports both formats.
+
+
+```java
+@Configuration
+@EnableAsync
+public class LightningConfiguration {
+  @Bean
+  public Executor taskExecutor() {
+    ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+    executor.setCorePoolSize(4);
+    executor.setMaxPoolSize(4);
+    executor.setQueueCapacity(2);
+    executor.setThreadNamePrefix("Lightning-");
+    executor.initialize();
+    return executor;
+  }
+
+  @Bean
+  public PipelineOptions pipelineOptions() {
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options.setRunner(FlinkRunner.class);
+    return options;
+  }
+}
+```
+
+Flink streaming pipelines in Beam are blocking operations so they will need to be run in a separate thread. To do so I'll create an Executor Bean with Lightning prefix for its Threads.
+
+In this configuration class I'll also define a Bean with Apache Beam's PipelineOptions that can be used across all the application. I will set just one parameter to use an embedded FlinkRunner. It starts Flink Master and job server automatically.
```

Comment:
> job server automatically

Jeśli nie wyjaśniasz, czym jest job server, to ja bym nic nie pisał o nim. Job server jest traktowany jako "implementation detail". Użytkownik Beama ma nie być świadomy jego istnienia.

Reply:
Prawda, zwłaszcza że potem podaję link do poczytania na temat Flinka w Beamie